### DISCUSSION TITLE
NextJS Monolith to serve directories under /pages as different projects with dynamic basePaths for a possible Nginx reverse proxy

### DISCUSSION THREAD

**Discussion by sumodhnair19** [vercel/next.js#44706] [Help]
I have a `NextJS` application which acts like a `monolith`. Means, I have multiple directories under `/pages` which acts as different projects.

     /pages/first_project/[...slugs]
     /pages/second_project/[...slugs]

For the setup I have Docker which runs `nextjs` as a container alongside `Nginx` as a prima-facie to reverse proxy back to the application. 
I would like `Nginx` to map its `/location` against the pages directory in NextJS. 

For e.g 

    /first_project --> /pages/first_project/
    /second_project --> /pages/second_project/

In my company, the hosting of the projects are like this:

    <company.com>/first_project/[...slugs]
    <company.com>/second_project/[...slugs]

The issue I am facing is, since `/location` in the `Nginx` looks for all the build files in that particular location. I can't have dynamic hosting possible. I was wondering if this is at all possible?

If I don't provide a `basePath` in `next.config.js`, then for `<company.com>/first_project/`, `Nginx` expects files to be available in the `location --> first_project`. However, files are available at the `root directory` of the project. So I end up getting error.  

I obviously can fix this issue by setting `basePath` in `next.config` as `first_project`. However, this will fail for `second_project`. Is there any way to dynamically load `basePath`? Please let me know.

Company Nginx Config:

    server {
        listen 80;
        server_name _;
        location /first_project {
          proxy_pass http://local:8000;
    
        }
        location /second_project { 
          proxy_pass http://local:8000;
    
        }
     
    }

(local:8000)NextJS Nginx Monolith Config

    upstream nextjs_upstream {
      server <nextjs_project>:3000;
    }
    
    server {
      listen 80 default_server;
    
      server_name _;
    
      server_tokens off;
    
      gzip on;
      gzip_proxied any;
      gzip_comp_level 4;
      gzip_types text/css application/javascript image/svg+xml;
    
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection 'upgrade';
      proxy_set_header Host $host;
      proxy_cache_bypass $http_upgrade;
    
      location /_next/static {

        proxy_pass http://nextjs_upstream;
      }

    
      location / {
        proxy_pass http://nextjs_upstream;
      }
    }

Any help will be appreciated. Thanks :)

**Comment by GabenGar** [comment]
Why should it even be a monolith in the first place? 2 potentially different "projects" will likely have different config files and therefore bundle output. You should revert the logic and instead use 2 separate nextjs projects with something like [`multi zones`](https://nextjs.org/docs/advanced-features/multi-zones) and nginx redirecting to a particular app as per path rules.

**Reply by sumodhnair19** [reply]
Hi @GabenGar Thanks for your response. I understand what you're saying. However, I have many components, auth etc I'd like to share between projects, and considering dynamic routing aspect of NextJS, I figured going with this approach. However, I have now stumbled onto `turborepo` that seems to fit my case very well. However, I'd need to go deeper around Next-Auth with it.

**Reply by GabenGar** [reply]
Yeah for shared code you'd have to use the likes of [npm workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces).
